toolchain/check/scope_stack.h

```diff
@@ -0,0 +1,195 @@
+            SemIR::NameScopeId scope_id = SemIR::NameScopeId::Invalid,
+            bool lexical_lookup_has_load_error = false) -> void;
+
+  // Pops the top scope from scope_stack_, cleaning up names from
+  // lexical_lookup_.
+  auto Pop() -> void;
+
+  // Pops the top scope from scope_stack_ if it contains no names.
+  auto PopIfEmpty() -> void {
+    if (scope_stack_.back().names.empty()) {
+      Pop();
+    }
+  }
+
+  // Pops scopes until we return to the specified scope index.
+  auto PopToScope(ScopeIndex index) -> void;
+
+  // Looks up the name `name_id` in the current scope. Returns the existing
+  // lookup result, if any.
+  auto LookupInCurrentScope(SemIR::NameId name_id) -> SemIR::InstId;
+
+  // Looks up the name `name_id` in the current scope and its enclosing scopes.
+  // Returns the innermost lexical lookup result, if any, along with a list of
+  // non-lexical scopes in which lookup should also be performed, ordered from
+  // outermost to innermost.
+  auto LookupInEnclosingScopes(SemIR::NameId name_id)
+      -> std::pair<SemIR::InstId, llvm::ArrayRef<NonLexicalScope>>;
+
+  // Looks up the name `name_id` in the current scope. Returns the existing
+  // instruction if any, and otherwise adds the name with the value `target_id`
+  // and returns Invalid.
+  auto LookupOrAddName(SemIR::NameId name_id, SemIR::InstId target_id)
+      -> SemIR::InstId;
+
+  // Returns the scope index associated with the current scope.
+  auto current_scope_index() const -> ScopeIndex {
+    return current_scope().index;
+  }
+
+  // Returns the name scope associated with the current lexical scope, if any.
+  auto current_scope_id() const -> SemIR::NameScopeId {
+    return current_scope().scope_id;
+  }
+
+  // Returns the instruction associated with the current scope, or Invalid if
+  // there is no such instruction, such as for a block scope.
+  auto current_scope_inst_id() const -> SemIR::InstId {
+    return current_scope().scope_inst_id;
+  }
+
+  // Returns the current scope, if it is of the specified kind. Otherwise,
+  // returns nullopt.
+  template <typename InstT>
+  auto GetCurrentScopeAs(const SemIR::File& sem_ir) -> std::optional<InstT> {
+    auto inst_id = current_scope_inst_id();
+    if (!inst_id.is_valid()) {
+      return std::nullopt;
+    }
+    return sem_ir.insts().TryGetAs<InstT>(inst_id);
+  }
+
+  // If there is no `returned var` in scope, sets the given instruction to be
+  // the current `returned var` and returns an invalid instruction ID. If there
+  // is already a `returned var`, returns it instead.
+  auto SetReturnedVarOrGetExisting(SemIR::InstId inst_id) -> SemIR::InstId;
+
+  // Runs verification that the processing cleanly finished.
+  auto VerifyOnFinish() -> void;
+
+  auto return_scope_stack() -> llvm::SmallVector<ReturnScope>& {
+    return return_scope_stack_;
+  }
+
+  auto break_continue_stack() -> llvm::SmallVector<BreakContinueScope>& {
+    return break_continue_stack_;
+  }
```

Comment:
Had you considered leaving these on Context? Versus the main scope stack, the SmallVector return means this doesn't really control either, and they're directly exposed on Context itself. You might need a pointer to update returned_var, but I don't even see logic for break_continue_stack.

Reply:
Yeah, I started that way, but the `returned var` handling seemed to fit better on `ScopeStack`, and the return scope stack is mixed up with that too. I was also thinking that we might actually want to switch to maintaining these stacks more like the non-lexical scope stacks, so that when you push a scope you say whether it's a return / break / continue scope, and don't have to remember to update these stacks when you enter a different scope. It probably doesn't matter right now that we're still in an enclosing return/break/continue scope when we enter, say, a nested class, but it could matter if we started allowing statements in subexpressions (eg, block expressions of some kind).